clarity/contracts/arkadiko-oracle-v1-1.clar

```diff
@@ -43,3 +53,102 @@
+  )
+    (if (>= check-result (var-get minimum-valid-signers))
+      (begin
+        (map-set prices { token: token-name } { last-price: price, last-block: block-height, decimals: decimals })
+        (ok true)
+      )
+      (ok false)
+    )
+  )
+)
+
+;; ---------------------------------------------------------
+;; Message signing
+;; ---------------------------------------------------------
+
+(define-read-only (pubkey-price-signer (block uint) (token-id uint) (price uint) (decimals uint) (signature (buff 65)))
+  (secp256k1-recover? (get-signable-message-hash block token-id price decimals) signature)
+)
+
+(define-read-only (check-price-signer (block uint) (token-id uint) (price uint) (decimals uint) (signature (buff 65)))
+  (let (
+    (pubKey (unwrap! (pubkey-price-signer block token-id price decimals signature) u0))
+  )
+    (if (is-trusted-oracle pubKey) u1 u0)
+  )
+)
+
+(define-read-only (get-signable-message-hash (block uint) (token-id uint) (price uint) (decimals uint))
+  (keccak256 (concat (concat (concat (uint256-to-buff-be block) (uint256-to-buff-be token-id)) (uint256-to-buff-be price)) (uint256-to-buff-be decimals)))
+)
+
+(define-read-only (uint256-to-buff-be (n uint))
+	(unwrap-panic (as-max-len? (get a (fold uint-to-buff-iter 0x0000000000000000000000000000000000000000000000000000000000000000 {n: n, a: 0x})) u32))
+)
+
+(define-constant byte-list 0x000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f202122232425262728292a2b2c2d2e2f303132333435363738393a3b3c3d3e3f404142434445464748494a4b4c4d4e4f505152535455565758595a5b5c5d5e5f606162636465666768696a6b6c6d6e6f707172737475767778797a7b7c7d7e7f808182838485868788898a8b8c8d8e8f909192939495969798999a9b9c9d9e9fa0a1a2a3a4a5a6a7a8a9aaabacadaeafb0b1b2b3b4b5b6b7b8b9babbbcbdbebfc0c1c2c3c4c5c6c7c8c9cacbcccdcecfd0d1d2d3d4d5d6d7d8d9dadbdcdddedfe0e1e2e3e4e5e6e7e8e9eaebecedeeeff0f1f2f3f4f5f6f7f8f9fafbfcfdfeff)
+
+(define-private (uint-to-buff-iter (b (buff 1)) (p {n: uint, a: (buff 32)}))
+	{
+		a: (unwrap-panic (as-max-len? (concat (if (is-eq (get n p) u0) 0x00 (unwrap-panic (element-at byte-list (mod (get n p) u256)))) (get a p)) u32)),
+		n: (/ (get n p) u256)
+	}
+)
+
+;; ---------------------------------------------------------
+;; Admin
+;; ---------------------------------------------------------
+
+(define-public (set-token-id (token-id uint) (token-name (string-ascii 12)))
+  (begin
+    (asserts! (is-eq tx-sender (contract-call? .arkadiko-dao get-dao-owner)) (err ERR-NOT-AUTHORIZED))
+
+    (map-set token-id-to-name token-id token-name)
+    (map-set token-name-to-id token-name token-id)
+    (ok true)
+  )
+)
+
+(define-public (set-minimum-valid-signers (minimum uint))
+  (begin
+    (asserts! (is-eq tx-sender (contract-call? .arkadiko-dao get-dao-owner)) (err ERR-NOT-AUTHORIZED))
+    
+    (var-set minimum-valid-signers minimum)
+    (ok true)
+  )
+)
+
+(define-public (set-trusted-oracle (pubkey (buff 33)) (trusted bool))
```

Comment:
Q: could we remove this method?

I know this reduces flexibility significantly, but I would hardcode the pubkeys of trusted oracles on deployment and if we need to update them, we would update the oracle contract through governance. We could also do this in a v3 if we know this version works well

the reason I'm asking is because this defeats the whole purpose of not having a single point of failure.. if the DAO owner keys would somehow be compromised, our oracle is also screwed... and if we remove this method, we at least keep our oracle safe?

Reply:
The DAO owner keys are not stored an a server, unlike the oracle keys.

Not sure if this method really matters for now because:
- There is another method "update-price-owner" to allow the DAO to update prices in case something goes wrong. Would like to keep this method for now until this new system has proven itself. So if the DAO keys are compromised, it's easier for the attacker to use the method "update-price-owner".
- If someone would get his hands on the DAO keys, he can get all liquidity from swap and reserves and all tokens in that wallet. Not sure if at that point the oracle still matters :P